Migrate core logic of deletecontent and exportcontent to a util function

thesujai opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->
This refactors the exportcontent and exportcontent commands and its related tasks

- Moved the exportcontent core logic to export_content util function
- Updated the diskexport task to use export_content instead of calling exportcontent command
- Moved the deletecontent core logic to delete_content util function
- Updated deletechannel task to user delete_content


## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->
Fixes #13209 
Fixes #13208 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

- Try soft delete and force delete a content
- Export a channel contents to a file

thesujai committed:
move delete_metadata to content_delete util file

thesujai committed:
refactor deletecontent command and task to use util function for core logic

thesujai committed:
move core logic to util functions

thesujai committed:
remove dummy job check

thesujai committed:
update exportcontent tests

thesujai committed:
use export_content util in diskexport task

thesujai commented:
Hi @rtibbles, marking this as ready for review

MisRob commented:
Thanks @thesujai! We will review in the upcoming weeks. I wanted to note that review waiting times are still longer.

rtibbles reviewed (commented):
Code changes all make sense to me, following the same pattern as previously. Tests are still passing, which is an excellent sign! I'll mark this for QA review to finalize.

rtibbles commented:
For manual QA here, testing these workflows:

* Exporting a channel - this time the important part is exporting the resources themselves, so selecting a range of resources in a channel and confirming that all works.
* Deleting resources from a channel, and deleting an entire channel.

radinamatic commented:
I tested a variety of delete & export scenarios of resources and full channels, always re-checking the size of the `content` folder to make sure it changes at every step, and a part from some discrepancies in the reported size of resources on device in Kolibri, did not notice any issues or regressions. 

Example here would be that the initial size of resources from the GDL channel present on device was 20MB, but resulted in (20+9+4=~33) MBs when I was deleting separate folders:
Before | After
-- | --
<img width="961" alt="2025-03-21_03-19-11" src="https://github.com/user-attachments/assets/28c06957-3a8b-490c-8f3f-166a736b7e88" /> | <img width="961" alt="2025-03-21_03-31-51" src="https://github.com/user-attachments/assets/07aa700f-f4de-48cd-a518-e684a94ab1c2" />

 The above almost certainly is not result of the changes in this PR, and no issues were observed in the Learn UI, as the deleted resources were properly not displayed at every step.

However at one point I noticed some request errors in the console, that at first did not seem an issue for performing delete operations, until I bumped into a situation where it was impossible to re-import from a second local drive resources from a channel that was originally exported there in the previous step. Sequence was the following:

1. Import some resources from the HP Life FR channel from Studio into Kolibri.
2. Select a portion of those to export on the second drive.
3. Delete the whole HP Life FR channel from Kolibri.
4. Try to import it from the local drive to confirm that all previously exported ones are indeed present.

Result was a hard stop:

<img width="961" alt="2025-03-21_03-59-01" src="https://github.com/user-attachments/assets/6a326f63-6500-4a79-b896-7402ba12e85f" />

Note that all 3 channels that are reporting 404 request errors were previously deleted from Kolibri during this testing session.

[DB-logs.zip](https://github.com/user-attachments/files/19380545/DB-logs.zip)

[Full home folder](https://drive.google.com/file/d/1GOFA7V5XTw9QZUO8RuV84JUunEnfBPN2/view?usp=drive_link) (>500MB)

thesujai commented:
Hi @radinamatic, Thank you for this!
I went further and tried exporting some contents of a channel to a drive(from release 0.17.x branch) then again switched to this branch and tried exporting different contents from the same channel. Then I deleted the channel the entire channel and tried to import from my local drive again. I was able to import the exported contents from release branch and not from the current branch.

So this  probably is a regression with export-channel refactor from my previous PR or an issue with export-content refactor from this PR

PS: looks like an issue with export-channel, i reverted export_channel refactor in this branch and then tried every steps again, it worked as expected with no problems

rtibbles commented:
@thesujai if you can work out how to fix the issue here, that would be great - if not, I am happy to make a separate fix on develop so that we don't release this regression :)

thesujai commented:
@rtibbles I thought I could solve it but I couldn't. Sorry for confusing you with my previous comment, because this regression is not caused by any of the refactor of management commands(Just observed). This has been there since a long time
In release 0.17.x branch:
In this video:

1. In a fresh instance of kolibri, I import resources(china folder) from Touchable Earth en
2. Export them to a pendrive
3. Delete the entire channel
4. Re-import entire resources from pendrive
5. Try importing `russia folder`, but this time from studio
6. Then again export `russia folder` to drive(Note China folder was already exported) 
7. Delete the entire channel again
8. Try to import from drive again, I only see the `china folder`(Where is the `russia folder`

[Screencast from 21-03-25 09:39:38 PM IST.webm](https://github.com/user-attachments/assets/b4b503a1-05df-436c-9b85-5656b56d15ff)


After this, I empty the pendrive(note in my device also all the channels are deleted)(I haven't restarted my kolibri instance)
1. I again import some resources from Touchable earth
2. Export them to drive(selective exporting)
3. Delete the channel
4. Try to import back from the drive again, I observe resources that I haven't even imported(Why are they in my drive?)
5. Anyways I import them, the import is successful(But i dont see the channel in device)(Obviously because the resources never existed in my drive)

[Screencast from 21-03-25 09:53:36 PM IST.webm](https://github.com/user-attachments/assets/8abf3ef0-ce7f-4bb6-b8a7-b43d29f9f38f)

rtibbles commented:
OK, as this seems to be pre-existing, I will file as a separate issue including the context from both of these reports. Thank you!

rtibbles commented:
Have rebased here because of our updates to our required checks, then this should be good to merge.

rtibbles reviewed (approved):
We will open the observed, and pre-existing, behaviour issues as a new issue. This PR is good to merge.
